Sources/XCTest/Public/Asynchronous/XCTWaiter.swift

```diff
@@ -356,22 +349,13 @@ open class XCTWaiter {
 
 private extension XCTWaiter {
     func primitiveWait(using runLoop: RunLoop, duration timeout: TimeInterval) {
-        // The contract for `primitiveWait(for:)` explicitly allows waiting for a shorter period than requested
-        // by the `timeout` argument. Only run for a short time in case `cancelPrimitiveWait()` was called and
-        // issued `CFRunLoopStop` just before we reach this point.
-        let timeIntervalToRun = min(0.1, timeout)
-
-        // RunLoop.run(mode:before:) should have @discardableResult <rdar://problem/45371901>
-        _ = runLoop.run(mode: .default, before: Date(timeIntervalSinceNow: timeIntervalToRun))
+        self.runLoopSource = RunLoop._Source()
+        self.runLoop?._add(self.runLoopSource!, forMode: .default)
+        _ = self.runLoop?.run(until: .init(timeIntervalSinceNow: timeout))
     }
 
     func cancelPrimitiveWait() {
-        guard let runLoop = runLoop else { return }
-#if os(Windows)
-        runLoop._stop()
-#else
-        CFRunLoopStop(runLoop.getCFRunLoop())
-#endif
+        runLoopSource?.invalidate()
```

Comment:
This `cancelPrimitiveWait()` method is always called from the `XCTWaiter.subsystemQueue` DispatchQueue, but with the changes in `primitiveWait` above, the `runLoopSource` stored property will now be read here on that private queue but assigned in `primitiveWait` which is _not_ on that same queue.

So I think there could be a data race lurking here. The current code in `primitiveWait` takes the RunLoop as a function parameter, which is done intentionally because the _assignment_ of that RunLoop to the `runLoop` stored property is done earlier, inside a `DispatchQueue.sync {...}` block. I think we ought to take a similar approach here by creating and storing the `RunLoop._Source` instance earlier in the calling `wait(...)` function, and assign it to the stored properties inside that protected `.sync {...}` block. Then, accessing the `runLoopSource` stored property here in `cancelPrimitiveWait()` should become safe. We might even consider adding a

```swift
dispatchPrecondition(condition: .onQueue(XCTWaiter.subsystemQueue))
```

to it, like other on-queue funcs have, to help ensure this.

Reply:
Have made the changes. Thanks